### PR TITLE
fix(aws-events-targets): Consume IRestApi as target

### PR DESCRIPTION
**Pull request by AWS-MattB**
Fixes: aws#16423

When creating an Event Bridge target for an API Gateway, the
superinterface IRestApi should be consumed instead of the
concrete class RestApi.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-events-targets/lib/api-gateway.ts [hunk 1]
```diff
@@ -74,7 +74,7 @@ export interface ApiGatewayProps extends TargetBaseProps {
  */
 export class ApiGateway implements events.IRuleTarget {
 
-  constructor(public readonly restApi: api.RestApi, private readonly props?: ApiGatewayProps) {
+  constructor(public readonly restApi: api.IRestApi, private readonly props?: ApiGatewayProps) {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-events-targets/lib/api-gateway.ts:77]: The error is:

```
API elements with incompatible changes:
err  - PROP @aws-cdk/aws-events-targets.ApiGateway.restApi: type @aws-cdk/aws-apigateway.IRestApi (formerly @aws-cdk/aws-apigateway.RestApi): @aws-cdk/aws-apigateway.IRestApi does not extend @aws-cdk/aws-apigateway.RestApi [changed-type:@aws-cdk/aws-events-targets.ApiGateway.restApi]
```

The reason why this change is incompatible is because `public readonly restApi` declares both an INPUT parameter and an OUTPUT parameter, and we can only change one in one direction, and the other in the other. That makes it somewhat more complex, so It probably needs to look somewhat like this:

```
private readonly _restApi: IRestApi;

constructor(restApi: IRestApi, private readonly props?: ApiGatewayProps) {
  this._restApi = resetApi;
}

public get restApi(): RestApi {
  if (!(this._restApi instanceof RestApi)) {
    throw new Error('The RestApi is not a RestApi construct, and cannot be retrieved this way');
  }
  return this._restApi;
}
```

**Review by rix0rrr — CHANGES_REQUESTED** [review]

**Comment by AWS-MattB** [comment]
Here are some possible alternatives, let me know what you prefer:

1. Add a `class SpecApiGateway implements events.IRuleTarget`, with a `public readonly specRestApi: api.SpecRestApi`, so that 100% of the known implementors of `api.IRestApi` are covered. Extract the common logic between the two to a private class
2. Add a non-abstract superclass of `ApiGateway` (open to naming suggestions) with a `public readonly iRestApi: api.IRestApi`, move most of the logic there. This would result in `ApiGateway` having two properties (`restApi`, `iRestApi`) referring to the same object, but would allow the gradual deprecation of `ApiGateway` in favor of the superclass.
3. Add an additional constructor which accepts `public readonly iRestApi: api.IRestApi` and explicitly define a getter method for `restApi: api.RestApi`, which throws an error if the `api.IRestApi` constructor was used (as you suggested above)

**Comment by AWS-MattB** [comment]
I've updated the PR to be option 2 (add a non-abstract superclass of ApiGateway) and have named it `ApiGatewayTarget`.

**Comment by AWS-MattB** [comment]
> ### AWS CodeBuild CI Report
> * CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
> * Commit ID: [4b31f64](https://github.com/aws/aws-cdk/commit/4b31f6424bcad13fc1d7f304adb15140720d29fc)
> * Result: FAILED
> * [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=5f6a35e9-2983-424f-b1cd-933ee00fd3ed%2Fbuild.log) (available for 30 days)
> 
> _Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)_

The failure is due to a 5xx error talking to Yarn's repo. This is a rebase of the above commit, which built OK in the CI, and this commit built OK on my local environment.

File: packages/@aws-cdk/aws-events-targets/lib/api-gateway.ts [hunk 2]
```diff
@@ -70,11 +70,14 @@ export interface ApiGatewayProps extends TargetBaseProps {
 }
 
 /**
- * Use an API Gateway REST APIs as a target for Amazon EventBridge rules.
+ * Use an API Gateway REST API as a target for Amazon EventBridge rules.
  */
-export class ApiGateway implements events.IRuleTarget {
+export class ApiGatewayTarget implements events.IRuleTarget {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-events-targets/lib/api-gateway.ts:75]: As far as I can tell, none of the other classes end in `Target`. 

I'm not sure this one should buck the trend.

File: packages/@aws-cdk/aws-events-targets/lib/api-gateway.ts [hunk 3]
```diff
@@ -70,11 +70,14 @@ export interface ApiGatewayProps extends TargetBaseProps {
...
 
-  constructor(public readonly restApi: api.RestApi, private readonly props?: ApiGatewayProps) {
+  /**
+   * @param iRestApi - An implementation of a Rest API to send events to
+   */
+  constructor(public readonly iRestApi: api.IRestApi, protected readonly props?: ApiGatewayProps) {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-events-targets/lib/api-gateway.ts:80]:
```suggestion
  constructor(public readonly restApi: api.IRestApi, protected readonly props?: ApiGatewayProps) {
```

File: packages/@aws-cdk/aws-events-targets/lib/api-gateway.ts [hunk 4]
```diff
@@ -121,3 +124,19 @@ export class ApiGateway implements events.IRuleTarget {
 
 }
 
+/**
+ * Use an API Gateway REST APIs as a target for Amazon EventBridge rules.
+ *
+ * @deprecated - Use ApiGatewayTarget
+ */
+export class ApiGateway extends ApiGatewayTarget {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-events-targets/lib/api-gateway.ts:132]: Oh I see. No, we can repurpose this class.

**Review by rix0rrr — CHANGES_REQUESTED** [review]
I think I told you what to do in the previous review comment.

Did it not work? If so, you should tell me about it. I think I prefer that solution.
